File: geekclaw/plugin/transport.go
package plugin

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"sync"
	"sync/atomic"
)

// Transport 处理通过 io.Reader/Writer 对（通常是外部进程的 stdio 管道）
// 进行的 JSON-RPC 2.0 双向通信。
//
// 支持三种消息流：
//   - Go→Plugin 请求/响应（Call）
//   - Plugin→Go 通知（Notifications）
//   - Plugin→Go 反向请求/响应（RegisterService + handleReverseCall）
type Transport struct {
	writer  io.Writer
	scanner *bufio.Scanner

	nextID  atomic.Int64
	pending sync.Map // id -> chan *Response

	// 从外部进程接收的通知
	notifyCh chan *Notification

	// 反向调用：插件通过 host.* 方法调用 Go 服务
	services sync.Map // method -> ServiceHandler
	ctx      context.Context

	mu sync.Mutex // 序列化写入
}

// NewTransport 创建一个基于给定 reader（进程 stdout）和
// writer（进程 stdin）的 Transport。
func NewTransport(r io.Reader, w io.Writer) *Transport {
	scanner := bufio.NewScanner(r)
	scanner.Buffer(make([]byte, 0, 1024*1024), 10*1024*1024) // 最大 10MB 行

	return &Transport{
		writer:   w,
		scanner:  scanner,
		notifyCh: make(chan *Notification, 64),
		ctx:      context.Background(),
	}
}

// SetContext 设置用于反向调用 ServiceHandler 的上下文。
// 必须在 ReadLoop 启动之前调用。
func (t *Transport) SetContext(ctx context.Context) {
	t.ctx = ctx
}

// RegisterService 注册一个 Go 服务处理函数，供插件通过反向 RPC 调用。
// method 应使用 "host.<service>.<action>" 格式。
func (t *Transport) RegisterService(method string, handler ServiceHandler) {
	t.services.Store(method, handler)
}

// Notifications 返回入站通知的只读通道。
func (t *Transport) Notifications() <-chan *Notification {
	return t.notifyCh
}

// Call 发送 JSON-RPC 请求并等待响应。
func (t *Transport) Call(ctx context.Context, method string, params any) (json.RawMessage, error) {
	id := t.nextID.Add(1)

	req := Request{
		JSONRPC: "2.0",
		ID:      id,
		Method:  method,
		Params:  params,
	}

	ch := make(chan *Response, 1)
	t.pending.Store(id, ch)
	defer t.pending.Delete(id)

	if err := t.send(req); err != nil {
		return nil, fmt.Errorf("send %s: %w", method, err)
	}

	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case resp := <-ch:
		if resp.Error != nil {
			return nil, resp.Error
		}
		raw, err := json.Marshal(resp.Result)
		if err != nil {
			return nil, fmt.Errorf("marshal result: %w", err)
		}
		return raw, nil
	}
}

// Notify 发送 JSON-RPC 通知（不期望响应）。
func (t *Transport) Notify(method string, params any) error {
	n := Notification{
		JSONRPC: "2.0",
		Method:  method,
		Params:  params,
	}
	return t.send(n)
}

// ReadLoop 从外部进程读取消息并分发。
// 阻塞直到 reader 关闭或发生错误。
//
// 三路分发：
//   - id != nil && method == "": 响应（Go 发出的请求的回复）
//   - id != nil && method != "": 反向调用（插件调用 Go 服务）
//   - id == nil && method != "": 通知（插件发出的单向消息）
func (t *Transport) ReadLoop() error {
	for t.scanner.Scan() {
		line := t.scanner.Bytes()
		if len(line) == 0 {
			continue
		}

		var peek struct {
			ID     *int64          `json:"id"`
			Method string          `json:"method"`
			Result json.RawMessage `json:"result"`
			Error  *RPCError       `json:"error"`
			Params json.RawMessage `json:"params"`
		}
		if err := json.Unmarshal(line, &peek); err != nil {
			continue
		}

		if peek.ID != nil && peek.Method == "" {
			// 响应：Plugin 回复 Go 发出的请求
			resp := &Response{
				JSONRPC: "2.0",
				ID:      *peek.ID,
				Error:   peek.Error,
			}
			if peek.Result != nil {
				var result any
				_ = json.Unmarshal(peek.Result, &result)
				resp.Result = result
			}
			if ch, ok := t.pending.Load(*peek.ID); ok {
				ch.(chan *Response) <- resp
			}
		} else if peek.ID != nil && peek.Method != "" {
			// 反向调用：Plugin 调用 Go 服务
			id := *peek.ID
			method := peek.Method
			params := peek.Params
			go t.handleReverseCall(id, method, params)
		} else if peek.Method != "" {
			// 通知
			var notif Notification
			if err := json.Unmarshal(line, &notif); err != nil {
				continue
			}
			select {
			case t.notifyCh <- &notif:
			default:
				// 缓冲区满时丢弃
			}
		}
	}

	return t.scanner.Err()
}

// handleReverseCall 处理来自插件的反向 JSON-RPC 调用。
func (t *Transport) handleReverseCall(id int64, method string, rawParams json.RawMessage) {
	handler, ok := t.services.Load(method)
	if !ok {
		_ = t.send(Response{
			JSONRPC: "2.0",
			ID:      id,
			Error:   &RPCError{Code: -32601, Message: fmt.Sprintf("method not found: %s", method)},
		})
		return
	}

	result, err := handler.(ServiceHandler)(t.ctx, rawParams)
	if err != nil {
		_ = t.send(Response{
			JSONRPC: "2.0",
			ID:      id,
			Error:   &RPCError{Code: -32000, Message: err.Error()},
		})
		return
	}

	_ = t.send(Response{
		JSONRPC: "2.0",
		ID:      id,
		Result:  result,
	})
}

// send 序列化并写入单行 JSON。
func (t *Transport) send(v any) error {
	data, err := json.Marshal(v)
	if err != nil {
		return err
	}
	data = append(data, '\n')

	t.mu.Lock()
	defer t.mu.Unlock()
	_, err = t.writer.Write(data)
	return err
}

File: crates/plugin/src/transport.rs
//! JSON-RPC 2.0 stdio 传输层。
//!
//! 处理通过子进程 stdin/stdout 管道进行的双向 JSON-RPC 通信。
//! 支持三种消息流：
//!   - 宿主→插件 请求/响应（call）
//!   - 插件→宿主 通知（notifications）
//!   - 插件→宿主 反向请求/响应（register_service + handle_reverse_call）

use std::collections::HashMap;
use std::sync::atomic::{AtomicI64, Ordering};
use std::sync::Arc;

use serde_json::Value;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::sync::{mpsc, oneshot, Mutex, RwLock};

use crate::error::{PluginError, Result};
use crate::wire::{
    JsonRpcNotification, JsonRpcRequest, JsonRpcResponse, PeekMessage, RpcError,
};

/// 服务处理函数类型：处理来自插件的反向 JSON-RPC 调用。
/// 插件通过 host.* 方法调用宿主服务。
pub type ServiceHandler = Box<dyn Fn(Value) -> std::pin::Pin<Box<dyn std::future::Future<Output = std::result::Result<Value, String>> + Send>> + Send + Sync>;

/// JSON-RPC 2.0 传输层。
///
/// 管理与子进程的双向通信：
///   - 向 stdin 写入请求
///   - 从 stdout 读取响应/通知/反向调用
///   - 原子请求 ID 计数器
///   - 待处理请求映射表
pub struct Transport {
    /// 子进程 stdin 写入端，序列化写入需要互斥
    writer: Arc<Mutex<tokio::process::ChildStdin>>,

    /// 原子请求 ID 计数器
    next_id: AtomicI64,

    /// 待处理请求：id -> 响应发送端
    pending: Arc<RwLock<HashMap<i64, oneshot::Sender<Result<Value>>>>>,

    /// 通知接收通道
    notification_rx: Mutex<mpsc::Receiver<JsonRpcNotification>>,

    /// 通知发送通道（供读取循环使用）
    notification_tx: mpsc::Sender<JsonRpcNotification>,

    /// 反向调用服务处理函数
    services: Arc<RwLock<HashMap<String, ServiceHandler>>>,
}

impl Transport {
    /// 创建新的传输层。
    ///
    /// `stdin` - 子进程 stdin 句柄（用于写入请求）
    ///
    /// 返回 (Transport, stdout_reader) 元组，调用者需要将 stdout_reader
    /// 传递给 `read_loop` 方法启动后台读取。
    pub fn new(stdin: tokio::process::ChildStdin) -> Self {
        let (tx, rx) = mpsc::channel(64);
        Self {
            writer: Arc::new(Mutex::new(stdin)),
            next_id: AtomicI64::new(0),
            pending: Arc::new(RwLock::new(HashMap::new())),
            notification_rx: Mutex::new(rx),
            notification_tx: tx,
            services: Arc::new(RwLock::new(HashMap::new())),
        }
    }

    /// 注册反向调用服务处理函数。
    /// method 应使用 "host.<service>.<action>" 格式。
    pub async fn register_service(&self, method: impl Into<String>, handler: ServiceHandler) {
        self.services.write().await.insert(method.into(), handler);
    }

    /// 发送 JSON-RPC 请求并等待响应。
    pub async fn call(&self, method: &str, params: Option<Value>) -> Result<Value> {
        let id = self.next_id.fetch_add(1, Ordering::Relaxed) + 1;

        let req = JsonRpcRequest::new(id, method, params);

        let (tx, rx) = oneshot::channel();
        self.pending.write().await.insert(id, tx);

        // 发送请求
        if let Err(e) = self.send_raw(&req).await {
            self.pending.write().await.remove(&id);
            return Err(e);
        }

        // 等待响应
        match rx.await {
            Ok(result) => result,
            Err(_) => Err(PluginError::ChannelClosed),
        }
    }

    /// 发送 JSON-RPC 通知（不期望响应）。
    pub async fn notify(&self, method: &str, params: Option<Value>) -> Result<()> {
        let notif = JsonRpcNotification {
            jsonrpc: "2.0".to_string(),
            method: method.to_string(),
            params,
        };
        self.send_raw(&notif).await
    }

    /// 获取通知接收器。
    /// 注意：同一时刻只能有一个调用者持有此锁。
    pub async fn recv_notification(&self) -> Option<JsonRpcNotification> {
        self.notification_rx.lock().await.recv().await
    }

    /// 从子进程 stdout 持续读取消息并分发。
    /// 阻塞直到 reader 关闭或发生错误。
    ///
    /// 三路分发：
    ///   - id 有值 && method 为空: 响应
    ///   - id 有值 && method 非空: 反向调用
    ///   - id 无值 && method 非空: 通知
    pub async fn read_loop(&self, stdout: tokio::process::ChildStdout) -> Result<()> {
        let reader = BufReader::new(stdout);
        let mut lines = reader.lines();

        while let Ok(Some(line)) = lines.next_line().await {
            if line.is_empty() {
                continue;
            }

            let peek: PeekMessage = match serde_json::from_str(&line) {
                Ok(p) => p,
                Err(_) => continue, // 无法解析的行直接跳过
            };

            if peek.id.is_some() && peek.method.is_empty() {
                // 响应：插件回复宿主发出的请求
                let id = peek.id.unwrap();
                if let Some(sender) = self.pending.write().await.remove(&id) {
                    let result = if let Some(err) = peek.error {
                        Err(PluginError::from(err))
                    } else {
                        Ok(peek.result.unwrap_or(Value::Null))
                    };
                    let _ = sender.send(result);
                }
            } else if peek.id.is_some() && !peek.method.is_empty() {
                // 反向调用：插件调用宿主服务
                let id = peek.id.unwrap();
                let method = peek.method;
                let params = peek.params.unwrap_or(Value::Null);
                self.handle_reverse_call(id, &method, params).await;
            } else if !peek.method.is_empty() {
                // 通知
                let notif: JsonRpcNotification = match serde_json::from_str(&line) {
                    Ok(n) => n,
                    Err(_) => continue,
                };
                // 缓冲区满时丢弃
                let _ = self.notification_tx.try_send(notif);
            }
        }

        Ok(())
    }

    /// 处理来自插件的反向 JSON-RPC 调用。
    async fn handle_reverse_call(&self, id: i64, method: &str, params: Value) {
        let services = self.services.read().await;
        let handler = services.get(method);

        let response = match handler {
            None => {
                JsonRpcResponse::error(id, RpcError::method_not_found(method))
            }
            Some(handler) => {
                match handler(params).await {
                    Ok(result) => JsonRpcResponse::success(id, result),
                    Err(msg) => JsonRpcResponse::error(id, RpcError::internal(msg)),
                }
            }
        };

        let _ = self.send_raw(&response).await;
    }

    /// 序列化并写入单行 JSON。
    async fn send_raw<T: serde::Serialize>(&self, msg: &T) -> Result<()> {
        let mut data = serde_json::to_vec(msg)?;
        data.push(b'\n');

        let mut writer = self.writer.lock().await;
        writer.write_all(&data).await?;
        writer.flush().await?;
        Ok(())
    }
}
